src/frontend/src/components/modal.ts

```diff
@@ -0,0 +1,59 @@
+import { html, TemplateResult, render } from "lit-html";
+
+export const createModal = ({
+  slot,
+  id,
+  confirmText = "Ok",
+}: {
+  slot: TemplateResult;
+  id?: string;
+  confirmText?: string;
+}): HTMLDialogElement => {
+  const modalHtml = html`
+    <form method="dialog" class="c-modal__content c-card c-card--modal">
+      <div class="c-modal__inner">
+        <span class="c-modal__close" data-closemodal>&times;</span>
+        ${slot}
+        <div class="c-modal__footer">
+          <button type="submit" class="c-button c-button--primary">
+            ${confirmText}
+          </button>
+        </div>
+      </div>
+    </form>
+  `;
+
+  const $modal = document.createElement("dialog");
+  $modal.id = id != null ? id : "modal";
+  $modal.classList.add("c-modal");
+  $modal.setAttribute("aria-modal", "true");
+
+  render(modalHtml, $modal);
+  return $modal;
+};
+
+export const modal = (slot: TemplateResult): Promise<void> => {
+  // create modal and append it to the body
+  const $modal = createModal({ slot });
+  document.body.appendChild($modal);
+
+  return new Promise((resolve) => {
```

Comment:
just an idea: instead of keeping a promise open, why not return a callback instead?

Reply:
that's because of me, I've asked @meodai to do that so that we can await the modal (i.e. wait for the user to ack)